trace/src/otel.rs

```diff
@@ -0,0 +1,369 @@
+use std::borrow::Cow;
+use std::future::Future;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use futures::{
+    future::{BoxFuture, Shared},
+    FutureExt, TryFutureExt,
+};
+use tokio::sync::mpsc;
+use tokio::task::JoinError;
+use tokio_util::sync::CancellationToken;
+
+use observability_deps::{
+    opentelemetry::sdk::export::trace::{ExportResult, SpanData, SpanExporter},
+    tracing::{error, info, warn},
+};
+
+use crate::ctx::{SpanContext, SpanId, TraceId};
+use crate::span::{MetaValue, SpanEvent, SpanStatus};
+use crate::{span::Span, TraceCollector};
+
+/// Size of the exporter buffer
+const CHANNEL_SIZE: usize = 1000;
+
+/// Maximum number of events that can be associated with a span
+const MAX_EVENTS: u32 = 10;
+
+/// Maximum number of attributes that can be associated with a span
+const MAX_ATTRIBUTES: u32 = 100;
+
+/// `OtelExporter` wraps a opentelemetry SpanExporter and sinks spans to it
+///
+/// In order to do this it spawns a background worker that pulls messages
+/// of a queue and writes them to opentelemetry. If this worker cannot keep
+/// up, and this queue fills up, spans will be dropped and warnings logged
+#[derive(Debug)]
+pub struct OtelExporter {
+    join: Shared<BoxFuture<'static, Result<(), Arc<JoinError>>>>,
+
+    sender: tokio::sync::mpsc::Sender<SpanData>,
+
+    shutdown: CancellationToken,
+}
+
+impl OtelExporter {
+    /// Creates a new `OtelExporter`
+    pub fn new<T: SpanExporter + 'static>(exporter: T) -> Self {
+        let shutdown = CancellationToken::new();
+        let (sender, receiver) = mpsc::channel(CHANNEL_SIZE);
+
+        let handle = tokio::spawn(background_worker(shutdown.clone(), exporter, receiver));
+        let join = handle.map_err(Arc::new).boxed().shared();
+
+        Self {
+            join,
+            shutdown,
+            sender,
+        }
+    }
+
+    /// Triggers shutdown of this `OtelExporter`
+    pub fn shutdown(&self) {
+        info!("jaeger exporter shutting down");
```

Comment:
```suggestion
        info!("otel exporter shutting down");
```

It is confusing to me to intermix the terms jaeger and otel. I don't really care which is chosen but I think they should be consistent

Reply:
Definitely, this thing use to be called JaegerExporter and I renamed it, I must have missed these :+1: